{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "from main import extract_rally_output, high_avail_fld, all_in_one_fld, \\\n",
    "    get_experiment_data_bars"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "def error_to_human(error_list):\n",
    "    if len(error_list) == 0:\n",
    "        return None\n",
    "    if re.compile('Quota exceeded for instances').search(error_list[2]):\n",
    "        return \"NOVA: Quota exceeded for instance\"\n",
    "    if re.compile('executer._rebuild_server').search(error_list[2]):\n",
    "        return \"NOVA: Rebuild server error\"\n",
    "    if re.compile('Quota exceeded.*router').search(error_list[2]):\n",
    "        return \"NEUTRON: Quota exceeded for router\"\n",
    "    if re.compile('Quota exceeded.*security_group').search(error_list[2]):\n",
    "        return \"NEUTRON: Quota exceeded for security group\"\n",
    "    if re.compile(' Maximum number of volumes allowed \\(10\\) exceeded for quota').search(error_list[2]):\n",
    "        return \"CINDER: Quota exceeded for volume\"\n",
    "    if re.compile('Unable to establish connection to http://130.149.249').search(error_list[2]):\n",
    "        #return \"CONNECTION_ERROR: Unreachable node,\"\n",
    "        if \"admin_keystone.create_user()\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Node is unreachable create user\"\n",
    "        if \"add_role(user_id=self.executer.user.id\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, add role\"\n",
    "        if \"_create_security_group\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, security group\"\n",
    "        if \"revoke_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, revoke role\"\n",
    "        if \"delete_user\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, delete user\"\n",
    "        if \"delete_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, delete role\"\n",
    "        if \"detach_volume\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, detach volume\"\n",
    "        if \"create_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, create role\"\n",
    "        if \"attach_volume\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, attach volume\"\n",
    "        return \"CONNECTION_ERROR: Unreachable node,\"\n",
    "    if re.compile('http://130.149.249.*timed out').search(error_list[2]):\n",
    "        #return \"CONNECTION_ERROR: Unreachable node,\"\n",
    "        if \"admin_keystone.create_user()\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, create user\"\n",
    "        if \"add_role(user_id=self.executer.user.id\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, add role\"\n",
    "        if \"_create_security_group\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, security group\"\n",
    "        if \"revoke_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, revoke role\"\n",
    "        if \"delete_user\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, delete user\"\n",
    "        if \"delete_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, delete role\"\n",
    "        if \"detach_volume\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, detach volume\"\n",
    "        if \"create_role\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, create role\"\n",
    "        if \"attach_volume\" in error_list[2]:\n",
    "            return \"CONNECTION_ERROR: Unreachable node, attach volume\"\n",
    "        return \"CONNECTION_ERROR: Unreachable node,\"\n",
    "    if re.compile('self.executer.create_volume_params').search(error_list[2]):\n",
    "        return \"CINDER: Create volume\"\n",
    "    if re.compile('Invalid volume: Volume status must be').search(error_list[2]):\n",
    "        return \"CINDER: Invalid volume\"\n",
    "    if re.compile('Cannot.*detach.*while it is in task_state rebuild').search(error_list[2]):\n",
    "        return \"NOVA: Detach volume\"\n",
    "    if re.compile('Cannot.*detach.*while it is in vm_state error').search(error_list[2]):\n",
    "        return \"NOVA: Detach volume\"\n",
    "    if re.compile('self.executer.server_kwargs').search(error_list[2]):\n",
    "        return \"NOVA: Create server\"\n",
    "    if re.compile('download.cirros-cloud.net.*Network is unreachable').search(error_list[2]):\n",
    "        return \"NETWORK_ERROR: Unable to download image\"\n",
    "    if re.compile('cirros-0.3.5-x86_64-disk.img.*Connection refused').search(error_list[2]):\n",
    "        return \"NETWORK_ERROR: Unable to download image\"\n",
    "    if re.compile('cirros-0.3.5-x86_64-disk.img.*Connection timed out').search(error_list[2]):\n",
    "        return \"NETWORK_ERROR: Unable to download image\"\n",
    "    if re.compile('cirros-0.3.5-x86_64-disk.img.*No route to host').search(error_list[2]):\n",
    "        return \"NETWORK_ERROR: Unable to download image\"\n",
    "    print(error_list[2])\n",
    "    raise Exception(error_list)\n",
    "\n",
    "\n",
    "def get_error_info(task_data, concurrency, fail_point = None):\n",
    "    task_data = pd.DataFrame(task_data[0], columns= [\"timestamp\", \"duration\", \"error\", \"atomic_actions\"])\n",
    "    error = task_data['error']\n",
    "    error_formated = error.apply(lambda x: error_to_human(x))\n",
    "    error_info = {}\n",
    "    amount_of_errors_per_wl = error_formated.apply(lambda x: x == None)\n",
    "    failed_workloads = error_formated[~error_formated.isnull()]\n",
    "    amount_of_errors = failed_workloads.index.size\n",
    "    success_workloads = error_formated[error_formated.isnull()]\n",
    "    amount_of_success = success_workloads.index.size\n",
    "    last_success_index = success_workloads.index[-1]\n",
    "    if not fail_point:\n",
    "        fail_point = last_success_index\n",
    "\n",
    "    failed = False\n",
    "\n",
    "    first_error = None\n",
    "    first_error_index = None\n",
    "    last_error_before_fail = None\n",
    "    last_error_before_fail_index = None\n",
    "    error_amount_before_fail = None\n",
    "    success_amount_before_fail = None\n",
    "    first_error_after_fail = None\n",
    "    first_error_after_fail_index = None\n",
    "    last_error_amount_before_fail = None\n",
    "\n",
    "    if amount_of_errors > 0:\n",
    "        if last_success_index+concurrency <= failed_workloads.index[-1]:\n",
    "            failed = True\n",
    "    errors_before_fail = None\n",
    "    errors_after_fail = None\n",
    "    if failed_workloads.size > 0:\n",
    "        first_error_index = failed_workloads.index[0]\n",
    "        first_error = failed_workloads[first_error_index]\n",
    "    if failed:\n",
    "        errors_before_fail_index = failed_workloads.index[failed_workloads.index <= fail_point]\n",
    "        errors_before_fail = failed_workloads[errors_before_fail_index]\n",
    "        error[errors_before_fail_index].apply(lambda x: error_to_human(x))\n",
    "        errors_after_fail_index = failed_workloads.index[failed_workloads.index > fail_point]\n",
    "        errors_after_fail = failed_workloads[errors_after_fail_index]\n",
    "\n",
    "        last_error_before_fail_index = errors_before_fail_index[-1]\n",
    "        last_error_before_fail = failed_workloads[last_error_before_fail_index]\n",
    "        last_error_amount_before_fail = errors_before_fail[errors_before_fail==last_error_before_fail].count()\n",
    "\n",
    "        error_amount_before_fail = errors_before_fail_index.size\n",
    "        success_amount_before_fail = success_workloads.size\n",
    "\n",
    "        first_error_after_fail_index = failed_workloads.index[failed_workloads.index > fail_point][0]\n",
    "        first_error_after_fail = failed_workloads[first_error_after_fail_index]\n",
    "\n",
    "    error_info.update({'first_error':first_error, 'first_error_index':first_error_index })\n",
    "    error_info.update({'last_error_before_fail':last_error_before_fail, \"last_error_before_fail_index\":last_error_before_fail_index})\n",
    "    error_info.update({'last_error_amount_before_fail': last_error_amount_before_fail})\n",
    "    error_info.update({\"error_amount_before_fail\":error_amount_before_fail })\n",
    "    error_info.update({\"success_amount_before_fail\":success_amount_before_fail})\n",
    "    error_info.update({\"first_error_after_fail\":first_error_after_fail, \"first_error_after_fail_index\":first_error_after_fail_index})\n",
    "\n",
    "    unique_errors = error_formated.dropna().unique()\n",
    "    error_stat = pd.DataFrame()\n",
    "    error_data = list()\n",
    "    for unique_error in unique_errors:\n",
    "        before_occ = None\n",
    "        before_occ_perc = None\n",
    "        after_occ = None\n",
    "        after_occ_perc = None\n",
    "        total_occ = error_formated[error_formated == unique_error].count()\n",
    "        total_occ_perc = (total_occ/error_formated.count()) * 100\n",
    "        if failed:\n",
    "            before_occ = errors_before_fail[errors_before_fail==unique_error].count()\n",
    "            before_occ_perc = (before_occ/errors_before_fail.count())*100\n",
    "            after_occ = errors_after_fail[errors_after_fail==unique_error].count()\n",
    "            after_occ_perc = (after_occ/errors_after_fail.count())*100\n",
    "        curr_stat = {\"Total occurence\":total_occ, \"Total occurence perc.\":total_occ_perc,\n",
    "                     \"Before failed occurence\": before_occ, \"Before failed occurence perc.\": before_occ_perc,\n",
    "                     \"After failed occurence\": after_occ, \"After failed occurence perc.\": after_occ_perc}\n",
    "        error_stat[unique_error] = curr_stat\n",
    "    if failed:\n",
    "        error_data = [errors_before_fail, errors_after_fail]\n",
    "    else:\n",
    "        error_data = [failed_workloads]\n",
    "    return error_info, error_stat, error_data"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "concurrency_list = [1,2,4,8,16,64]\n",
    "AiO_iteration_to_use=[1,4,1,4,2,1]\n",
    "AiO_failure_point = [None, None, None, 1064, 11, 2620]\n",
    "HA_error_info = pd.DataFrame()\n",
    "AiO_error_info = pd.DataFrame()\n",
    "HA_error_stats = list()\n",
    "AiO_error_stats = list()\n",
    "HA_error_data = list()\n",
    "AiO_error_data = list()\n",
    "cached = False\n",
    "if cached:\n",
    "    HA_error_info = pd.read_csv(\"tmp/HA_error.csv\", index_col=0)\n",
    "    AiO_error_info = pd.read_csv(\"tmp/AiO_error.csv\", index_col=0)\n",
    "    for i in range(6):\n",
    "        HA_error_stats.append(pd.read_csv(f\"tmp/HA_stat{i}.csv\", index_col=0))\n",
    "        AiO_error_stats.append(pd.read_csv(f\"tmp/AiO_stat{i}.csv\", index_col=0))\n",
    "else:\n",
    "    for scenario_counter in range(6):\n",
    "        conc = concurrency_list[scenario_counter]\n",
    "        HA_data = extract_rally_output(high_avail_fld, conc)\n",
    "        AiO_data = extract_rally_output(all_in_one_fld, conc, AiO_iteration_to_use[scenario_counter])\n",
    "        error_info, error_stats, error_data = get_error_info(HA_data, conc)\n",
    "        HA_error_info = pd.concat([HA_error_info, pd.DataFrame([error_info])], ignore_index=True)\n",
    "        HA_error_stats.append(error_stats)\n",
    "        HA_error_data.append(error_data)\n",
    "        error_info, error_stats, error_data = get_error_info(AiO_data, conc, AiO_failure_point[scenario_counter])\n",
    "        AiO_error_info = pd.concat([AiO_error_info, pd.DataFrame([error_info])], ignore_index=True)\n",
    "        AiO_error_stats.append(error_stats)\n",
    "        AiO_error_data.append(error_data)\n",
    "\n",
    "    if not os.path.exists(\"tmp\"):\n",
    "        os.mkdir(\"tmp\")\n",
    "    HA_error_info.to_csv(\"tmp/HA_error.csv\")\n",
    "    AiO_error_info.to_csv(\"tmp/AiO_error.csv\")\n",
    "    for index in range(6):\n",
    "        HA_error_stats[index].to_csv(f\"tmp/HA_stat{index}.csv\")\n",
    "        AiO_error_stats[index].to_csv(f\"tmp/AiO_stat{index}.csv\")\n",
    "config_list = [\"Mult\", \"AiO\"]\n",
    "concurrencies=[1,2,4,8,16,64]\n",
    "total_data = pd.DataFrame()\n",
    "for i, data in enumerate(HA_error_stats):\n",
    "    if i > 2:\n",
    "        total_data = total_data.add(data, fill_value=0)\n",
    "for i, data in enumerate(AiO_error_stats):\n",
    "    if i > 2:\n",
    "        total_data = total_data.add(data, fill_value=0)\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "exp6_bars = get_experiment_data_bars(high_avail_fld, 64, 1, 20)\n",
    "HA_bars = list()\n",
    "AiO_bars = list()\n",
    "for i in range(6):\n",
    "    HA_bars.append(get_experiment_data_bars(high_avail_fld, concurrency_list[i], 1, 20))\n",
    "    AiO_bars.append(get_experiment_data_bars(all_in_one_fld, concurrency_list[i], AiO_iteration_to_use[i], 20))"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "plt.rcParams['font.size'] = 16\n",
    "plt.rcParams['figure.figsize'][0] = 12"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "({'first_error': 'NEUTRON: Quota exceeded for security group', 'first_error_index': 9, 'last_error_before_fail': 'NEUTRON: Quota exceeded for security group', 'last_error_before_fail_index': 43213, 'last_error_amount_before_fail': 40509, 'error_amount_before_fail': 40704, 'success_amount_before_fail': 2511, 'first_error_after_fail': 'NEUTRON: Quota exceeded for security group', 'first_error_after_fail_index': 43215},                                NEUTRON: Quota exceeded for security group   \n",
      "Total occurence                                              94340.000000  \\\n",
      "Total occurence perc.                                           91.537133   \n",
      "Before failed occurence                                      40509.000000   \n",
      "Before failed occurence perc.                                   99.520932   \n",
      "After failed occurence                                       53831.000000   \n",
      "After failed occurence perc.                                    86.325732   \n",
      "\n",
      "                               NEUTRON: Quota exceeded for router   \n",
      "Total occurence                                         83.000000  \\\n",
      "Total occurence perc.                                    0.080534   \n",
      "Before failed occurence                                 72.000000   \n",
      "Before failed occurence perc.                            0.176887   \n",
      "After failed occurence                                  11.000000   \n",
      "After failed occurence perc.                             0.017640   \n",
      "\n",
      "                               CINDER: Create volume   \n",
      "Total occurence                             1.000000  \\\n",
      "Total occurence perc.                       0.000970   \n",
      "Before failed occurence                     1.000000   \n",
      "Before failed occurence perc.               0.002457   \n",
      "After failed occurence                      0.000000   \n",
      "After failed occurence perc.                0.000000   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, add role   \n",
      "Total occurence                                                    9.000000  \\\n",
      "Total occurence perc.                                              0.008733   \n",
      "Before failed occurence                                            4.000000   \n",
      "Before failed occurence perc.                                      0.009827   \n",
      "After failed occurence                                             5.000000   \n",
      "After failed occurence perc.                                       0.008018   \n",
      "\n",
      "                               CONNECTION_ERROR: Node is unreachable create user   \n",
      "Total occurence                                                       138.000000  \\\n",
      "Total occurence perc.                                                   0.133900   \n",
      "Before failed occurence                                                44.000000   \n",
      "Before failed occurence perc.                                           0.108097   \n",
      "After failed occurence                                                 94.000000   \n",
      "After failed occurence perc.                                            0.150742   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, security group   \n",
      "Total occurence                                                         2.000000   \\\n",
      "Total occurence perc.                                                   0.001941    \n",
      "Before failed occurence                                                 2.000000    \n",
      "Before failed occurence perc.                                           0.004914    \n",
      "After failed occurence                                                  0.000000    \n",
      "After failed occurence perc.                                            0.000000    \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, revoke role   \n",
      "Total occurence                                                       6.000000  \\\n",
      "Total occurence perc.                                                 0.005822   \n",
      "Before failed occurence                                               4.000000   \n",
      "Before failed occurence perc.                                         0.009827   \n",
      "After failed occurence                                                2.000000   \n",
      "After failed occurence perc.                                          0.003207   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, delete user   \n",
      "Total occurence                                                       8.000000  \\\n",
      "Total occurence perc.                                                 0.007762   \n",
      "Before failed occurence                                               3.000000   \n",
      "Before failed occurence perc.                                         0.007370   \n",
      "After failed occurence                                                5.000000   \n",
      "After failed occurence perc.                                          0.008018   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, delete role   \n",
      "Total occurence                                                       7.000000  \\\n",
      "Total occurence perc.                                                 0.006792   \n",
      "Before failed occurence                                               1.000000   \n",
      "Before failed occurence perc.                                         0.002457   \n",
      "After failed occurence                                                6.000000   \n",
      "After failed occurence perc.                                          0.009622   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, detach volume   \n",
      "Total occurence                                                         2.000000  \\\n",
      "Total occurence perc.                                                   0.001941   \n",
      "Before failed occurence                                                 2.000000   \n",
      "Before failed occurence perc.                                           0.004914   \n",
      "After failed occurence                                                  0.000000   \n",
      "After failed occurence perc.                                            0.000000   \n",
      "\n",
      "                               NOVA: Quota exceeded for instance   \n",
      "Total occurence                                      8413.000000  \\\n",
      "Total occurence perc.                                   8.163047   \n",
      "Before failed occurence                                25.000000   \n",
      "Before failed occurence perc.                           0.061419   \n",
      "After failed occurence                               8388.000000   \n",
      "After failed occurence perc.                           13.451361   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, create role   \n",
      "Total occurence                                                      21.000000  \\\n",
      "Total occurence perc.                                                 0.020376   \n",
      "Before failed occurence                                               6.000000   \n",
      "Before failed occurence perc.                                         0.014741   \n",
      "After failed occurence                                               15.000000   \n",
      "After failed occurence perc.                                          0.024055   \n",
      "\n",
      "                               CONNECTION_ERROR: Unreachable node, attach volume   \n",
      "Total occurence                                                         3.000000  \\\n",
      "Total occurence perc.                                                   0.002911   \n",
      "Before failed occurence                                                 3.000000   \n",
      "Before failed occurence perc.                                           0.007370   \n",
      "After failed occurence                                                  0.000000   \n",
      "After failed occurence perc.                                            0.000000   \n",
      "\n",
      "                               CINDER: Quota exceeded for volume   \n",
      "Total occurence                                        24.000000  \\\n",
      "Total occurence perc.                                   0.023287   \n",
      "Before failed occurence                                24.000000   \n",
      "Before failed occurence perc.                           0.058962   \n",
      "After failed occurence                                  0.000000   \n",
      "After failed occurence perc.                            0.000000   \n",
      "\n",
      "                               NOVA: Create server  \n",
      "Total occurence                           5.000000  \n",
      "Total occurence perc.                     0.004851  \n",
      "Before failed occurence                   4.000000  \n",
      "Before failed occurence perc.             0.009827  \n",
      "After failed occurence                    1.000000  \n",
      "After failed occurence perc.              0.001604  , [9        NEUTRON: Quota exceeded for security group\n",
      "10       NEUTRON: Quota exceeded for security group\n",
      "11       NEUTRON: Quota exceeded for security group\n",
      "12       NEUTRON: Quota exceeded for security group\n",
      "13       NEUTRON: Quota exceeded for security group\n",
      "                            ...                    \n",
      "43209    NEUTRON: Quota exceeded for security group\n",
      "43210    NEUTRON: Quota exceeded for security group\n",
      "43211    NEUTRON: Quota exceeded for security group\n",
      "43212    NEUTRON: Quota exceeded for security group\n",
      "43213    NEUTRON: Quota exceeded for security group\n",
      "Name: error, Length: 40704, dtype: object, 43215     NEUTRON: Quota exceeded for security group\n",
      "43216     NEUTRON: Quota exceeded for security group\n",
      "43217     NEUTRON: Quota exceeded for security group\n",
      "43218                            NOVA: Create server\n",
      "43219              NOVA: Quota exceeded for instance\n",
      "                             ...                    \n",
      "105568    NEUTRON: Quota exceeded for security group\n",
      "105569    NEUTRON: Quota exceeded for security group\n",
      "105570    NEUTRON: Quota exceeded for security group\n",
      "105571    NEUTRON: Quota exceeded for security group\n",
      "105572    NEUTRON: Quota exceeded for security group\n",
      "Name: error, Length: 62358, dtype: object])\n"
     ]
    }
   ],
   "source": [
    "HA_data = extract_rally_output(high_avail_fld, 64)\n",
    "temp = get_error_info(HA_data, 64)\n",
    "print (temp)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [
    "def plot_error_data(bars, error, save_file=None):\n",
    "    concurrency_list = [1, 2, 4, 8, 16, 64]\n",
    "    AiO_iteration_to_use = [1, 4, 1, 4, 2, 1]\n",
    "\n",
    "    error = error[0]\n",
    "    unique_errors = error.dropna().unique()\n",
    "    c = list()\n",
    "    for er in unique_errors:\n",
    "        if \"Quota exceeded for security\" not in er:\n",
    "            c.append(er)\n",
    "    unique_errors = c\n",
    "\n",
    "    bar_error = {}\n",
    "\n",
    "    for unique_error in unique_errors:\n",
    "        bar_error[unique_error] = list()\n",
    "\n",
    "    for bar_index in bars.index:\n",
    "        start_index = bars['start_index'][bar_index]\n",
    "        end_index = bars['end_index'][bar_index]\n",
    "        errors_in_bar = error.iloc[(error.index>=start_index) & (error.index<= end_index)]\n",
    "        for unique_error in unique_errors:\n",
    "            bar_error[unique_error].append(int(errors_in_bar[errors_in_bar==unique_error].count()))\n",
    "    y_offset = np.zeros(bars['hour'].size)\n",
    "    for unique_error in unique_errors:\n",
    "        if sum(bar_error[unique_error]) == 0:\n",
    "            continue\n",
    "        label = unique_error.split(\":\")[0]+\"\\n\"+unique_error.split(\":\")[1][1:]\n",
    "        plt.bar(x=bars['hour']*5, height= bar_error[unique_error], bottom=y_offset, label=label, width=4)\n",
    "        plt.ylabel(\"Amount of errors per 5 minutes\")\n",
    "        plt.xlabel(\"Experiment duration (minute)\")\n",
    "        y_offset = y_offset + bar_error[unique_error]\n",
    "    plt.legend()\n",
    "    plt.grid()\n",
    "    plt.tight_layout()\n",
    "    plt.ylim((0,35))\n",
    "    plt.xlim((0,100))\n",
    "    if save_file:\n",
    "        fig = plt.gcf()\n",
    "        fig.savefig(f'fig/{save_file}', dpi=fig.dpi)\n",
    "    plt.show()\n",
    "\n",
    "def plot_error_data_all(HA_error_data, AiO_error_data, HA_bars, AiO_bars):\n",
    "    plot_error_data(AiO_bars[5][0][0][0:24], AiO_error_data[5], \"AiO_64_errors.pdf\")\n",
    "    return\n",
    "    plot_error_data(HA_bars[5][0][0], HA_error_data[5], \"HA_64_errors.pdf\")\n",
    "    for i in range (6):\n",
    "        plot_error_data(HA_bars[i][0][0], HA_error_data[i], None)\n",
    "    for i in range (6):\n",
    "        plot_error_data(AiO_bars[i][0][0], AiO_error_data[i], None)\n",
    "\n",
    "plot_error_data_all(HA_error_data, AiO_error_data, HA_bars, AiO_bars)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}